File: 2023/elixir/day02b/day02b.exs
defmodule Day02a do
  @progname "day02a.exs"

  def usage do
    IO.puts("usage: #{@progname} <file>")
    System.halt(1)
  end

  def process(contents) do
    lines = String.split(contents, "\n", trim: true)
    Enum.reduce(lines, 0, fn line, result ->
      draws_part = hd(tl(String.split(line, ": ")))
      {red_needed, green_needed, blue_needed} = Enum.reduce(String.split(draws_part, "; "), {0, 0, 0}, fn draws, {red, green, blue} ->
        Enum.reduce(String.split(draws, ", "), {red, green, blue}, fn draw, {current_red, current_green, current_blue} ->
          amount = String.to_integer(hd(String.split(draw, " ")))
          color = hd(tl(String.split(draw, " ")))
          case color do
            "red" ->
              if amount > current_red do
                {amount, current_green, current_blue}
              else
                {current_red, current_green, current_blue}
              end
            "green" ->
              if amount > current_green do
                {current_red, amount, current_blue}
              else
                {current_red, current_green, current_blue}
              end
            "blue" ->
              if amount > current_blue do
                {current_red, current_green, amount}
              else
                {current_red, current_green, current_blue}
              end
          end
        end)
      end)
      result + red_needed * green_needed * blue_needed
    end)
  end

  def main do
    if length(System.argv()) < 1 do
      usage()
    end

    filename = hd(System.argv())
    {_, contents} = File.read(filename)
    result = process(contents)
    IO.puts("result = #{result}")
  end
end

Day02a.main()

File: 2023/erlang/day02b/day02b.erl
-module (day02b).
-export ([main/1]).

usage() ->
    Module = "day02b.erl",
    io:format("usage: escript ~p <file>~n", [Module]),
    halt(1).

process_lines([], Result) -> Result;
process_lines([Line | Rest], Result) ->
    {RedNeeded, GreenNeeded, BlueNeeded} = process_line(Line),
    process_lines(Rest, Result + RedNeeded * GreenNeeded * BlueNeeded).

process_line(Line) ->
    DrawsPart = lists:last(string:split(Line, ": ")),
    {RedNeeded, GreenNeeded, BlueNeeded} = lists:foldl(
        fun(Draws, {CurrentRed, CurrentGreen, CurrentBlue}) ->
            lists:foldl(
                fun(ColorAmount, {RedNeeded, GreenNeeded, BlueNeeded}) ->
                    [AmountStr | _] = string:split(ColorAmount, " "),
                    Color = lists:last(string:split(ColorAmount, " ")),
                    {Amount, _} = string:to_integer(AmountStr),
                    case Color of
                        "red" ->
                            if Amount > RedNeeded ->
                                {Amount, GreenNeeded, BlueNeeded};
                            true ->
                                {RedNeeded, GreenNeeded, BlueNeeded}
                            end;
                        "green" ->
                            if Amount > GreenNeeded ->
                                {RedNeeded, Amount, BlueNeeded};
                            true ->
                                {RedNeeded, GreenNeeded, BlueNeeded}
                            end;
                        "blue" ->
                            if Amount > BlueNeeded ->
                                {RedNeeded, GreenNeeded, Amount};
                            true ->
                                {RedNeeded, GreenNeeded, BlueNeeded}
                            end;
                        _ -> throw({error, "unexpected color"})
                    end
                end,
                {CurrentRed, CurrentGreen, CurrentBlue}, string:split(Draws, ", ", all)
            )
        end,
        {0, 0, 0}, string:split(DrawsPart, "; ", all)
    ),
    {RedNeeded, GreenNeeded, BlueNeeded}.

process_input(Contents) ->
    Trimmed = string:trim(Contents),
    Lines = string:split(Trimmed, "\n", all),
    process_lines(Lines, 0).

main([Filename]) ->
    {ok, File} = file:read_file(Filename),
    Contents = unicode:characters_to_list(File),
    Result = process_input(Contents),
    io:format("result = ~w~n", [Result]);
main(_) ->
    usage().
